### PR TITLE
DF-390 Protect get by id method

### PR DESCRIPTION
**Pull request by varusakov**

### CODE REVIEW

File: application/src/main/java/ch/admin/seco/jobs/services/jobadservice/application/jobadvertisement/JobAdvertisementApplicationService.java [hunk 1]
```diff
@@ -282,6 +283,7 @@ public JobAdvertisementId updateFromX28(UpdateJobAdvertisementFromX28Dto updateJ
         );
     }
 
+    @PostAuthorize("isAuthenticated() || returnObject.status != T(ch.admin.seco.jobs.services.jobadservice.domain.jobadvertisement.JobAdvertisementStatus).PUBLISHED_RESTRICTED")
```
Comment (pavax) [inline application/src/main/java/ch/admin/seco/jobs/services/jobadservice/application/jobadvertisement/JobAdvertisementApplicationService.java]: I would do a PreAuthorize check and inject the JobAdvertisementAuthorizationService

```java
public canViewJob(JobAdvertisementId jobAdvertisementId){
// check that the JobAd is not in status PUBLISHED_RESTRICTED OR the Current User gas the Role ROLE_JOBSEEKER_CLIENT
}
```
```java
@PreAuthorize("@jobAdvertisementAuthorizationService.canViewJob(#id)")
```

File: application/src/main/java/ch/admin/seco/jobs/services/jobadservice/application/jobadvertisement/JobAdvertisementApplicationService.java [hunk 2]
```diff
@@ -282,6 +283,7 @@ public JobAdvertisementId updateFromX28(UpdateJobAdvertisementFromX28Dto updateJ
         );
     }
 
+    @PreAuthorize("@jobAdvertisementAuthorizationService.canViewJob(#jobAdvertisementId)")
```
[the remark this answers is not in the document]
Comment (varusakov) [reply application/src/main/java/ch/admin/seco/jobs/services/jobadservice/application/jobadvertisement/JobAdvertisementApplicationService.java]: @cjpsycos should we also protect getOneByFingerprint endpoint?

[the remark this answers is not in the document]
Comment (cjpsycos) [reply application/src/main/java/ch/admin/seco/jobs/services/jobadservice/application/jobadvertisement/JobAdvertisementApplicationService.java]: getByToken can only be used if the user has the token
getByFingerprint can only be used if it was send back from x28, and this is when the JobAd is in status PUBLUSHED_PUBLIC

so not needed

File: application/src/main/java/ch/admin/seco/jobs/services/jobadservice/application/jobadvertisement/JobAdvertisementAuthorizationService.java [hunk 3]
```diff
@@ -55,4 +58,21 @@ public boolean isOwner(JobAdvertisement jobAdvertisement, CurrentUser currentUse
...
         return (companyId != null) && companyId.equals(jobAdvertisement.getOwner().getCompanyId());
     }
+
+    public boolean canViewJob(JobAdvertisementId jobAdvertisementId) {
+        return this.jobAdvertisementRepository.findById(jobAdvertisementId)
+                .map(this::canViewJob)
+                .orElse(false);
+    }
+
+    public boolean canViewJob(String code) {
```
Comment (cjpsycos) [inline application/src/main/java/ch/admin/seco/jobs/services/jobadservice/application/jobadvertisement/JobAdvertisementAuthorizationService.java]: rename code --> stellennummer

File: domain/src/main/java/ch/admin/seco/jobs/services/jobadservice/domain/jobadvertisement/JobAdvertisementRepository.java [hunk 4]
```diff
@@ -44,6 +44,9 @@
 
     Optional<JobAdvertisement> findByFingerprint(String fingerprint);
 
+    @Query("select j from JobAdvertisement j where j.stellennummerAvam = :code or j.stellennummerEgov = :code")
+    Optional<JobAdvertisement> findByStellennummerAvamOrStellennummerEgov(String code);
```
Comment (cjpsycos) [inline domain/src/main/java/ch/admin/seco/jobs/services/jobadservice/domain/jobadvertisement/JobAdvertisementRepository.java]: rename code --> stellennummer

**Pull request merged**